Added ListTileControlAffinity into the ListTileTheme

abikko opened the pull request:
Added ListTileControlAffinity into the ListTileTheme. But we have an Design problem. In the ListTileTheme class we have overrided hashCode method. It uses Object.hash method, method accepts only 20 properties, but with a new controlAffinity property, it will be 21.

Issue: [Issue number one](https://github.com/flutter/flutter/issues/146414).

IMHO, no needs tests to change tests.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

abikko committed:
feat: Added controlAffinity property into the list_tile_theme.dart

Added control affinity property, also added it to copyWith, lerp, operator == and etc.

abikko committed:
added getting ListTileControlAffinity from theme if it not specified

If controlAffinity not specified to ListTileControlAffinity, it will try to get controlAffinity from ListTileTheme.of(context).controlAffinity, if not specified from the old code.

abikko committed:
Using Object.hashAll instead of Object.hash

abikko committed:
Added documentation for new property in list_tile_theme.dart

HansMuller commented on packages/flutter/lib/src/material/expansion_tile.dart:
```diff
@@ -669,14 +669,14 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
 
   Widget? _buildLeadingIcon(BuildContext context) {
-    if (_effectiveAffinity(widget.controlAffinity) != ListTileControlAffinity.leading) {
+    if (_effectiveAffinity(widget.controlAffinity ?? ListTileTheme.of(context).controlAffinity) != ListTileControlAffinity.leading) {
```
It looks like _effectiveAffinity could be `getEffectiveAffinity()`:
```dart
  ListTileControlAffinity _getEffectiveAffinity() {
    final ListTileControlAffinity affinity =
      widget.controlAffinity ?? ListTileTheme.of(context).controlAffinity ?? ListTileControlAffinity.trailing;
    return affinity == ListTileControlAffinity.platform ? ListTileControlAffinity.trailing : affinity;
  }
```

And then this expression could be `(_getEffectiveAffinity() != ListTileControlAffinity.leading)`

I just typed this in without really checking it. The idea is that the common code could be factored out and the effectiveAffinity computation could be simpler.

abikko replied:
resolved, see [commit](https://github.com/flutter/flutter/pull/146630/commits/eb31a043f8b4031cc63b3632221d98141c7aad3e)

HansMuller commented on packages/flutter/lib/src/material/expansion_tile.dart:
```diff
@@ -669,14 +669,14 @@ class _ExpansionTileState extends State<ExpansionTile> with SingleTickerProvider
 
   Widget? _buildTrailingIcon(BuildContext context) {
-    if (_effectiveAffinity(widget.controlAffinity) != ListTileControlAffinity.trailing) {
+    if (_effectiveAffinity(widget.controlAffinity ?? ListTileTheme.of(context).controlAffinity) != ListTileControlAffinity.trailing) {
```
See previous comment.

abikko replied:
resolved, see [commit](https://github.com/flutter/flutter/pull/146630/commits/eb31a043f8b4031cc63b3632221d98141c7aad3e)

abikko replied:
@TahaTesser I guess you should look at this?

HansMuller commented on packages/flutter/lib/src/material/list_tile_theme.dart:
```diff
@@ -462,6 +477,12 @@ class ListTileTheme extends InheritedTheme {
   /// [ListTileThemeData.enableFeedback] property instead.
   bool? get enableFeedback => _data != null ? _data.enableFeedback : _enableFeedback;
 
+  /// Overrides the default value of [ListTile.enableFeedback].
```
Looks like this comment was copied but not updated.

abikko replied:
resolved, see https://github.com/flutter/flutter/pull/146630/commits/550d9ed65a70d9afadbe3255d27d1035dbc9b688

HansMuller reviewed (commented):
Looks like a good start

HansMuller commented:
@TahaTesser - could you take over reviewing this PR?

TahaTesser commented on packages/flutter/lib/src/material/list_tile_theme.dart:
```diff
@@ -202,31 +208,34 @@ class ListTileThemeData with Diagnosticable {
+      enableFeedback,
+      mouseCursor,
+      visualDensity,
+      titleAlignment,
+    ],
```
```suggestion
      titleAlignment,
      controlAffinity,
    ],
```

TahaTesser commented on packages/flutter/lib/src/material/list_tile_theme.dart:
```diff
@@ -346,7 +358,8 @@ class ListTileTheme extends InheritedTheme {
        _horizontalTitleGap = horizontalTitleGap,
        _minVerticalPadding = minVerticalPadding,
-       _minLeadingWidth = minLeadingWidth;
+       _minLeadingWidth = minLeadingWidth,
+        _controlAffinity = controlAffinity;
```
```suggestion
       _minLeadingWidth = minLeadingWidth,
       _controlAffinity = controlAffinity;
```

TahaTesser reviewed (changes requested):
Thanks for the contribution!

As Hans said, this is a good start.  While you address the feedback this PR also need tests in order to land.  This makes sure the changes are working as expect and avoid breaking it in the future.

Commit:
Update packages/flutter/lib/src/material/list_tile_theme.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

Commit:
Update packages/flutter/lib/src/material/list_tile_theme.dart

Co-authored-by: Taha Tesser <tessertaha@gmail.com>

abikko commented:
@TahaTesser Yes, of course this update will need some tests.

But I have main work, and I am sometimes not able to write tests, I need an inspiration for it =)

But I will try to do as soon as possible for this tests

Commit:
refactoring

Renamed _effectiveAffinity method to _getEffectiveAffinity after review, also getting widget.controlAffinity instead of passing it as argument

Commit:
update docs

TahaTesser commented:
@abikko 
Thanks for addressing the suggestions. This still requires tests and also looks like CLA is missing.

Piinks commented:
Thank you for your contribution. I'm going to close this PR for now since there are outstanding comments and the CLA has not been signed, just to get this off our PR review queue. Please don't hesitate to re-open or submit a new PR if you have the time to address the review comments. Thanks!
